outlines/text/generate/regex.py

```diff
@@ -0,0 +1,160 @@
+
+    >>> import outlines.text as text
+    >>> sequence = text.generate.regex(model, "(0|[1-9][0-9]+)")("Return an integer between 0 and 10")
+
+    """
+
+    def __init__(self, model, regex_string: str, max_tokens: Optional[int]):
+        super().__init__(model, max_tokens)
+
+        vocabulary = model.tokenizer.vocabulary
+        sorted_vocabulary = [
+            k for k, v in sorted(vocabulary.items(), key=lambda kv: kv[1])
+        ]
+
+        regex_pattern = interegular.parse_pattern(regex_string)
+        self.regex_fsm = regex_pattern.to_fsm().reduce()
+
+        def partial_match_filter(string, end_idx, state_seq):
+            if end_idx is not None and end_idx < len(string) - 1:
+                return False
+            return True
+
+        pstate_to_vocab = map_partial_states_to_vocab(
+            list(sorted_vocabulary),
+            {"REGEX": self.regex_fsm},
+            partial_match_filter,
+            final_state_string=model.tokenizer.eos_token,
+        )
+
+        # TODO: This check might be a little too strict, because I think that
+        # while some states are made unreachable by a vocabulary (and will not
+        # be present in the following set difference), there could still be
+        # paths to terminal states emanating from the states that are reachable.
+        states_with_transition = {x[1] for x in pstate_to_vocab.keys()}
+        if len(self.regex_fsm.states.difference(states_with_transition)) > 0:
+            raise ValueError(
+                "The vocabulary does not allow us to build a sequence that matches the input regex"
+            )
+
+        self.pstate_to_vocab = {k: list(v) for k, v in pstate_to_vocab.items()}
+        # These tuples are comprised of the FSM name, last FSM state, and
+        # number of processed tokens.
+        # When an EOS is observed, the last FSM state becomes `-1`.
+        self.pstates: List[Tuple[str, int, int]] = []
+
+    def create_proposal(
+        self, generated_token_ids: torch.LongTensor, logits: torch.DoubleTensor
+    ) -> torch.DoubleTensor:
+        """Modify the next-token logits so that only integers can be generated.
+
+        Parameters
+        ----------
+        generated_token_ids
+            The token ids generated so far.
+        logits
+            The next-token logits.
+
+        """
+
+        if len(self.pstates) == 0:
+            self.pstates = [
+                ("REGEX", self.regex_fsm.initial, 0)
+                for _ in range(generated_token_ids.shape[0])
+            ]
+
+        if generated_token_ids.shape[-1] > 0:
+            new_pstates = []
+            for token_seq, (_, last_fsm_state, last_token_idx) in zip(
+                generated_token_ids,
+                self.pstates,
+            ):
+                # Get the tokens we haven't already processed
+                readable_tokens = token_seq[last_token_idx:]
+                # excluding any EOS tokens
+                not_eos_mask = [
```

Comment:
You should never get a sequence with an `EOS` token here. Those are filtered out in `Sequence.__call__`.

Reply:
If you start it out with a sequence like `[[10, 2, 0, 0]]`, you would only want to process the first two.  That's what it should be doing.

Reply:
When would that happen if you cannot get sequences with `0` by design?

Reply:
I believe I had to add it for the tests at one point, but that might not longer be true.